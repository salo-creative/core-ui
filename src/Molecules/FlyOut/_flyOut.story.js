import React from 'react';
import { storiesOf } from '@storybook/react';

// load tests
import { withTests } from '@storybook/addon-jest';
import results from '../../../.storybook/jest-test-results.json';

// FEATURED COMPONENTS //
import { FlyOut, FlyOutLink, FlyOutButton } from '../../index';

// README //
import README from './README.md';

// Start of story logic
const stories = storiesOf('Molecules | FlyOut', module);
stories.addDecorator(withTests({ results }));
stories.addParameters({ jest: ['flyout'] });

stories.add(
  'Basic',
  (() => {
    return (
      <FlyOut context='float'>
        <FlyOutButton title='Title' onClick={ () => alert('Button Clicked') } icon='sync' />
        <FlyOutLink title='Title' link='#' icon='dashboard' />
      </FlyOut>
    );
  }), { notes: README }
);