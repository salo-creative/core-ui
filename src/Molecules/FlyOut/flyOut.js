import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Icon from '@salo/icons';
import Ink from 'react-ink';

// COMPONENTS & STYLES
import { Container, Actions, Wrapper, Toggle } from './flyOut.styles';

// HELPERS & CONSTANTS
import { colours } from '../../helpers/colours';

class FlyOut extends Component {
  constructor(props) {
    super(props);
    this.state = { inlineMenu: false };
    this.inlineMenuContainer = null;
  }

  componentDidMount() {
    const { onGetRef } = this.props;

    if (typeof onGetRef === 'function') onGetRef(this.inlineMenuContainer);
  }

  toggleInlineMenu = () => {
    this.setState(prevState => ({ inlineMenu: !prevState.inlineMenu }));
  }

  handleClick(e) {
    const target = e.target.className;
    if (typeof target === 'string') {
      if (target.includes('ui-inline-menu__row')) {
        this.setState({ inlineMenu: false });
      }
    }
  }

  renderChildren() {
    const { children } = this.props;

    const childrenWithProps = React.Children.map(children, (child) => {
      if (child !== null) {
        return (
          React.cloneElement(child, { toggle: this.toggleInlineMenu })
        );
      }
      return null;
    }).filter(i => i);

    return childrenWithProps;
  }

  render() {
    const { mobile, context, inverse } = this.props;
    const { inlineMenu } = this.state;

    let setContext = context;
    if (context === 'table' && mobile) {
      setContext = 'table--card';
    }

    return (
      <Container
        className={ `salo-flyout ${ setContext } ${ inlineMenu ? 'is-active' : '' } ${ inverse ? 'inverse' : '' }` }
      >
        <Wrapper>
          <Actions
            ref={ (inlineMenuContainer) => { this.inlineMenuContainer = inlineMenuContainer; } }
          >
            { this.renderChildren() }
          </Actions>
          <Toggle onClick={ () => { this.toggleInlineMenu(); } }>
            <Ink />
            <Icon fill={ inverse ? colours.charcoal : '#fff' } icon='dots_vertical' size={ 24 } />
          </Toggle>
        </Wrapper>
      </Container>
    );
  }
}

FlyOut.defaultProps = {
  children: '',
  context: 'float',
  inverse: true,
  mobile: false,
  onGetRef: null
};

FlyOut.propTypes = {
  children: PropTypes.any,
  context: PropTypes.oneOf(['float', 'table', 'card']),
  inverse: PropTypes.bool,
  mobile: PropTypes.bool,
  onGetRef: PropTypes.func
};

export default FlyOut;