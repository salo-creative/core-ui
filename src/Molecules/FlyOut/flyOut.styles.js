import styled from 'styled-components';

export const Container = styled.div`
  z-index: 5;
  background: none;
  &:before {
    content: '';
    display: block;
    height: 0;
    width: 0;
    position: absolute;
    top: 0;
    left: 0;
    -webkit-transition: all 0.2s ease-in-out;
    transition: all 0.2s ease-in-out;
  }
  &.is-active:before {
    width: 100%;
    height: 100%;
    background: rgba(255, 255, 255, 0.98);
    &:not(.table) {
      height: calc(100% + 0.3rem);
    }
  }
  &:not(.card) {
    width: 100%;
    height: 100%;
  }
  /* Contexts */
  &.card {
    display: flex;
    justify-content: flex-end;
  }
  &.table--card {
    display: flex;
    justify-content: flex-end;
  }
  &.float,
  &.float-horizontal {
    position: relative;
  }
  &.table {
    align-items: center;
    display: flex;
  }
`;

export const Actions = styled.div`
  position: absolute;
  display: flex;
  width: 100%;
  visibility: hidden;
  justify-content: flex-end;

  .is-active & {
    visibility: visible;
  }

  /* Contexts */
  .card & {
    flex-direction: column;
    right: 0;
    bottom: 6rem;
    left: 0;
    padding: 0 0 0.5rem;
    width: calc(100% - 6rem);
    text-align: right;
  }

  .table & {
    flex-direction: row;
    right: 6rem;
    width: calc(100% - 6rem);
    padding: 0 1rem 0 0;
    top: 50%;
    transform: translateY(-50%);
    text-align: right;
  }

  .table--card & {
    flex-direction: column;
    right: 0;
    bottom: 4rem;
    left: 0;
    padding: 0 0 1rem;
    text-align: right;
  }

  .float & {
    flex-direction: row;
    justify-content: flex-start;
    left: 6.5rem;
    width: calc(100% - 6.5rem);
    top: 50%;
    transform: translateY(-50%);
    text-align: left;
  }

  .float-horizontal & {
    flex-direction: column;
    justify-content: inherit;
    left: 0;
    right: 0;
    text-align: right;
    top: 100%;
  }
`;

export const Wrapper = styled.div`
  display: inline-block;
  .card & {
    position: absolute;
    right: 1rem;
    bottom: 1rem;
  }
`;

export const Toggle = styled.button`
  cursor: pointer;
  display: flex;
  outline: none;
  border-radius: 50%;
  background: inherit;
  height: 4rem;
  width: 4rem;
  overflow: hidden;
  border: none;
  transform: rotate(90deg);
  transition: all 0.3s linear;
  &:focus,
  &:active {
    outline: none;
  }
  .is-active & {
    transition: transform 250ms;
    transform: rotate(0);
  }
  .inverse & {
    background: transparent;
    &:hover {
      background: ${ ({ theme }) => theme.grey };
    }
  }
`;

export const Item = styled.div`
  display: flex;
  justify-content: flex-end;
  margin: 0;
  transform: scale(0);
  transition-duration: .1s;
  .is-active & {
    transform: scale(1);
    transition-duration: .3s;
  }

  /* Contexts */
  .card & {
    padding-top: 0.5rem;
    padding-left: 0;
    padding-bottom: 0.5rem;
    padding-right: 0;
  }
  .table & {
    padding-top: 0;
    padding-bottom: 0;
  }
  .table--card & {
    padding-left: 0;
    padding-right: 0;
  }
  .float & {
    padding-top: 0;
    padding-bottom: 0;
  }
  .float-horizontal & {
    padding-left: 0;
    padding-right: 0;
  }
`;

export const ItemButton = styled.button`
  text-decoration: none;
  display: flex;
  align-items: center;
  -webkit-appearance: none;
  border: none;
  background: transparent;
  margin: 0;
  padding: 1rem;
  color: inherit;
  &:focus,
  &:active {
    outline: none;
  }
  .table & {
    flex-direction: row-reverse;
  }
`;

export const Title = styled.span`
  font-size: 1.2rem;
  text-transform: uppercase;
  font-weight: 500;
  letter-spacing: 1px;
  white-space: nowrap;
  color: inherit;
`;

export const IconWrapper = styled.div`
  cursor: pointer;
  outline: none;
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 4rem;
  background: #fff;
  height: 4rem;
  width: 4rem;
  border: 0.1rem solid inherit;
  display: flex;
  margin-left: 1rem;
  box-sizing: border-box;
  svg path {
    fill: inherit;
  }
  &:after {
    content: '';
    display: block;
    position: absolute;
    right: 0;
    top: 0;
    height: 4rem;
    border-radius: 4rem;
    width: 4rem;
  }
  .table & {
    margin-left: 0;
    margin-right: 1rem;
  }
`;