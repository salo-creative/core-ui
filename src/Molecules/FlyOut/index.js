import FlyOut from './flyOut';
import FlyOutButton from './flyOut.button';
import FlyOutLink from './flyOut.link';

export { FlyOutButton, FlyOutLink };

export default FlyOut;