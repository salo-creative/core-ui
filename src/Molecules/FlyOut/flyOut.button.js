import React from 'react';
import PropTypes from 'prop-types';
import Icon from '@salo/icons';

// COMPONENTS & STYLES
import { Item, ItemButton, Title, IconWrapper } from './flyOut.styles';

// HELPERS
import { colours } from '../../helpers/colours';

const FlyOutButton = (props) => {
  const { icon, onClick, title, toggle } = props;
  return (
    <Item>
      <ItemButton
        onClick={ (e) => onClick({ e, toggle }) }
        title={ title }
        className='salo-flyout__button'
      >
        <Title>{ title }</Title>
        <IconWrapper>
          <Icon fill={ colours.charcoal } icon={ icon } size={ 24 } />
        </IconWrapper>
      </ItemButton>
    </Item>
  );
};

FlyOutButton.propTypes = {
  title: PropTypes.string,
  icon: PropTypes.string,
  onClick: PropTypes.func.isRequired,
  toggle: PropTypes.func
};

FlyOutButton.defaultProps = {
  title: '',
  icon: 'dashboard',
  toggle: () => null
};

export default FlyOutButton;