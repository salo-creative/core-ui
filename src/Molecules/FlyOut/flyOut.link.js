import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import Icon from '@salo/icons';

// COMPONENTS & STYLES
import { Item, ItemButton, Title, IconWrapper } from './flyOut.styles';

// HELPERS
import { colours } from '../../helpers/colours';

const FlyOutLink = (props) => {
  const { icon, title, link } = props;
  return (
    <Item>
      <ItemButton
        to={ link }
        title={ title }
        as={ Link }
        className='salo-flyout__button'
      >
        <Title>{ title }</Title>
        <IconWrapper>
          <Icon fill={ colours.charcoal } icon={ icon } size={ 18 } />
        </IconWrapper>
      </ItemButton>
    </Item>
  );
};

FlyOutLink.propTypes = {
  title: PropTypes.string,
  icon: PropTypes.string,
  link: PropTypes.string
};

FlyOutLink.defaultProps = {
  title: '',
  icon: 'dashboard',
  link: ''
};

export default FlyOutLink;