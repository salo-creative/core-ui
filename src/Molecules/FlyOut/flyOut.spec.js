import React from 'react';
import 'jest-styled-components';

import { renderWithTheme } from '../../../test';

// COMPONENT
import FlyOut, { FlyOutLink } from './index';

test('Renders against snapshot', async () => {
  const { container } = renderWithTheme(
    <FlyOut>
      <FlyOutLink title='Title' link='#' icon='dashboard' />
    </FlyOut>
  );
  expect(container.firstChild).toMatchSnapshot();
});