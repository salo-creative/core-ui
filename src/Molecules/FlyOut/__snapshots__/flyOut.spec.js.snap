// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders against snapshot 1`] = `
.c7 {
  height: 18px;
  width: 18px;
  margin: 0;
}

.c9 {
  height: 24px;
  width: 24px;
  margin: 0;
}

.c0 {
  z-index: 5;
  background: none;
}

.c0:before {
  content: '';
  display: block;
  height: 0;
  width: 0;
  position: absolute;
  top: 0;
  left: 0;
  -webkit-transition: all 0.2s ease-in-out;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
}

.c0.is-active:before {
  width: 100%;
  height: 100%;
  background: rgba(255,255,255,0.98);
}

.c0.is-active:before:not(.table) {
  height: calc(100% + 0.3rem);
}

.c0:not(.card) {
  width: 100%;
  height: 100%;
}

.c0.card {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c0.table--card {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.c0.float,
.c0.float-horizontal {
  position: relative;
}

.c0.table {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c2 {
  position: absolute;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  visibility: hidden;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
}

.is-active .c2 {
  visibility: visible;
}

.card .c2 {
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  right: 0;
  bottom: 5rem;
  padding: 0 0 0.5rem;
  width: 100%;
  text-align: right;
}

.card .c2 .salo-flyout__button {
  padding: 0;
}

.table .c2 {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  right: 6rem;
  width: calc(100% - 6rem);
  padding: 0 1rem 0 0;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  text-align: right;
}

.table--card .c2 {
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  right: 0;
  bottom: 4rem;
  left: 0;
  padding: 0 0 1rem;
  text-align: right;
}

.float .c2 {
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  left: 6.5rem;
  width: calc(100% - 6.5rem);
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  text-align: left;
}

.float-horizontal .c2 {
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: inherit;
  -webkit-justify-content: inherit;
  -ms-flex-pack: inherit;
  justify-content: inherit;
  left: 0;
  right: 0;
  text-align: right;
  top: 100%;
}

.c1 {
  display: inline-block;
}

.card .c1 {
  position: absolute;
  z-index: 5;
  right: 1rem;
  bottom: 1rem;
}

.card__actions-top .c1 {
  top: 1rem;
}

.c8 {
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  outline: none;
  border-radius: 50%;
  background: inherit;
  height: 4rem;
  width: 4rem;
  overflow: hidden;
  border: none;
  -webkit-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
  -webkit-transition: all 0.3s linear;
  transition: all 0.3s linear;
}

.c8:focus,
.c8:active {
  outline: none;
}

.is-active .c8 {
  -webkit-transition: -webkit-transform 250ms;
  -webkit-transition: transform 250ms;
  transition: transform 250ms;
  -webkit-transform: rotate(0);
  -ms-transform: rotate(0);
  transform: rotate(0);
}

.inverse .c8 {
  background: transparent;
}

.inverse .c8:hover {
  background: #e9ecef;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  margin: 0;
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
  -webkit-transition-duration: .1s;
  transition-duration: .1s;
}

.is-active .c3 {
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
  -webkit-transition-duration: .3s;
  transition-duration: .3s;
}

.card .c3 {
  padding-top: 0.5rem;
  padding-left: 0;
  padding-bottom: 0.5rem;
  padding-right: 0;
}

.table .c3 {
  padding-top: 0;
  padding-bottom: 0;
}

.table--card .c3 {
  padding-left: 0;
  padding-right: 0;
}

.float .c3 {
  padding-top: 0;
  padding-bottom: 0;
}

.float-horizontal .c3 {
  padding-left: 0;
  padding-right: 0;
}

.c4 {
  -webkit-text-decoration: none;
  text-decoration: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-appearance: none;
  border: none;
  background: transparent;
  margin: 0;
  padding: 1rem;
  color: inherit;
}

.c4:focus,
.c4:active {
  outline: none;
}

.table .c4 {
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

.float-horizontal .c4 {
  padding-right: 0;
}

.c5 {
  font-size: 1.2rem;
  text-transform: uppercase;
  font-weight: 500;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  white-space: nowrap;
  color: inherit;
}

.c6 {
  cursor: pointer;
  outline: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 4rem;
  background: #fff;
  height: 4rem;
  width: 4rem;
  border: 0.1rem solid inherit;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-left: 1rem;
  box-sizing: border-box;
}

.c6 svg path {
  fill: inherit;
}

.c6:after {
  content: '';
  display: block;
  position: absolute;
  right: 0;
  top: 0;
  height: 4rem;
  border-radius: 4rem;
  width: 4rem;
}

.table .c6 {
  margin-left: 0;
  margin-right: 1rem;
}

<div
  class="c0 salo-flyout float  inverse "
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <a
          class="c4 salo-flyout__button"
          href="/"
          title="Title"
        >
          <span
            class="c5"
          >
            Title
          </span>
          <div
            class="c6 salo-flyout__icon-wrapper"
          >
            <svg
              class="c7"
              size="18"
              viewBox="0 0 24 24"
            >
              <path
                d="M13,3V9H21V3M13,21H21V11H13M3,21H11V15H3M3,13H11V3H3V13Z"
                fill="#333333"
              />
            </svg>
          </div>
        </a>
      </div>
    </div>
    <button
      class="c8 salo-flyout__toggle"
    >
      <canvas
        height="0"
        style="border-radius: inherit; height: 100%; left: 0px; position: absolute; top: 0px; width: 100%;"
        width="0"
      />
      <svg
        class="c9"
        size="24"
        viewBox="0 0 24 24"
      >
        <path
          d="M12,16A2,2 0 0,1 14,18A2,2 0 0,1 12,20A2,2 0 0,1 10,18A2,2 0 0,1 12,16M12,10A2,2 0 0,1 14,12A2,2 0 0,1 12,14A2,2 0 0,1 10,12A2,2 0 0,1 12,10M12,4A2,2 0 0,1 14,6A2,2 0 0,1 12,8A2,2 0 0,1 10,6A2,2 0 0,1 12,4Z"
          fill="#333333"
        />
      </svg>
    </button>
  </div>
</div>
`;
